### PR TITLE
More push based operations

### LINKED ISSUE #78: 1 second delay on .net framework when receiving messages
There is something funky going on on .net framework when receiving messages, when there is nothing buffered in the queue. From my testing there seems to be about 1 second delay between the message being send from service A, to it being received in service B, provided service B is a .net framework application. I have not observed the same behavior on .net core. 

Looking at the code in this library, the transport has a timeout of 2 seconds when receiving messages, which i what i suspect .net framework always waits, worst case scenario. I suspect it's the `SharedQueue` that doesn't notify the callling code that a new message has been received, causing the delay. 

I have not done any carefull measures, but looking at the traces emitted from Rebus.OpenTelemetry, it certainly seems to be the case. I have tested everything on the same host, including Rabbit, so it really shouldn't be a network issue.

### PR DESCRIPTION
**Pull request by zlepper**
# Basic introduction
I opened this in soon forever ago: https://github.com/rebus-org/Rebus/issues/917

I originally gave up, but I think I actually managed to make something that is more compatible with Rebus existing "pull" model, without requiring a change to the worker model also. 

This is just a draft, however I think it gives a basic for something we can talk about and discuss. Because it gets rid of the old `SharedQueue` class, it also fixes #78 as a side effect.

Do note the PR is a rough idea right now (Thus all the `Console.WriteLine` littered around in the code), but i hope it can serve as a starting pooint. 

## Odd behavior
There are however some very strange behavior that i have observed, however that i cannot quite explain, so i hope you can help me become smarter there. 

In the existing implemention (7.3.0 and the default thread based worker model) the following happens when I start my project:
```txt
[-] 27/04/2021 19.16.30 Info: Starting rebus {}
[-] 27/04/2021 19.16.30 Info: Queue started {}
[-] 27/04/2021 19.16.30 Info: Initializing RabbitMQ connection manager for transport with input queue "DebouncingService" {}
[-] 27/04/2021 19.16.30 Info: Mapped DebouncingService.Shared.Commands.DebounceMessage -> "DebouncingService" {}
[-] 27/04/2021 19.16.30 Info: Mapped DebouncingService.Shared.Commands.HandleDebouncedMessage -> "DebouncingService" {}
[-] 27/04/2021 19.16.30 Info: Mapped StatusService.Shared.Commands.RerunErroredJob -> "batch.status" {}
[-] 27/04/2021 19.16.30 Info: Mapped StatusService.Shared.Commands.SaveRetryableJob -> "batch.status" {}
[-] 27/04/2021 19.16.30 Info: Starting periodic task "CleanupTrackedErrors" with interval 00:00:10 {}
[-] 27/04/2021 19.16.30 Info: Bus "Rebus 1" started {}
[-] 27/04/2021 19.16.30 Info: Bus "Rebus 1" setting number of workers to 1 {}
[-] 27/04/2021 19.16.30 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.16.30 Info: Rebus started {}
```
Notice how only 1 consumer is initialized. 

Using the implementation in this PR, with the default thread based worker model, i get the following:
```txt
[-] 27/04/2021 19.18.59 Info: Queue started {}
[-] 27/04/2021 19.18.59 Info: Starting rebus {}
[-] 27/04/2021 19.18.59 Info: Initializing RabbitMQ connection manager for transport with input queue "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Mapped DebouncingService.Shared.Commands.DebounceMessage -> "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Mapped DebouncingService.Shared.Commands.HandleDebouncedMessage -> "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Mapped StatusService.Shared.Commands.RerunErroredJob -> "batch.status" {}
[-] 27/04/2021 19.18.59 Info: Mapped StatusService.Shared.Commands.SaveRetryableJob -> "batch.status" {}
[-] 27/04/2021 19.18.59 Info: Starting periodic task "CleanupTrackedErrors" with interval 00:00:10 {}
[-] 27/04/2021 19.18.59 Info: Bus "Rebus 1" started {}
[-] 27/04/2021 19.18.59 Info: Bus "Rebus 1" setting number of workers to 1 {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Rebus started {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.18.59 Info: Successfully initialized consumer for "DebouncingService" {}
```
Notice how there is considerably more than one consumer. To be specific i'm pretty sure there are 32 (Which is the number of cpu threads i have on this machine and is because I have `SetMaxParallelism(Environment.ProcessorCount)`. 

If I then which to the tpl based worker, i get the following:
```txt
[-] 27/04/2021 19.20.51 Info: Queue started {}
[-] 27/04/2021 19.20.51 Info: Starting rebus {}
[-] 27/04/2021 19.20.51 Info: Initializing RabbitMQ connection manager for transport with input queue "DebouncingService" {}
[-] 27/04/2021 19.20.51 Info: Mapped DebouncingService.Shared.Commands.DebounceMessage -> "DebouncingService" {}
[-] 27/04/2021 19.20.51 Info: Mapped DebouncingService.Shared.Commands.HandleDebouncedMessage -> "DebouncingService" {}
[-] 27/04/2021 19.20.51 Info: Mapped StatusService.Shared.Commands.RerunErroredJob -> "batch.status" {}
[-] 27/04/2021 19.20.51 Info: Mapped StatusService.Shared.Commands.SaveRetryableJob -> "batch.status" {}
[-] 27/04/2021 19.20.51 Info: Starting periodic task "CleanupTrackedErrors" with interval 00:00:10 {}
[-] 27/04/2021 19.20.51 Info: Bus "Rebus 1" started {}
[-] 27/04/2021 19.20.51 Info: Bus "Rebus 1" setting number of workers to 1 {}
[-] 27/04/2021 19.20.51 Info: Successfully initialized consumer for "DebouncingService" {}
[-] 27/04/2021 19.20.51 Info: Rebus started {}
```
And again, it only starts one consumer. 

I assume it's some of the extra async-await the thread model doesn't like, however that is a best guess. 

## Possible improvements. 
Currently it's possible for Rebus to manage to start many consumers (As clearly seen in the above section). However with the changes here it should be very easy to only ever have one consumer, and then have all threads read from that consumer. One worry i have about that is related to how the RabbitMQ client requires thread safety. AMQP is supposed to allow sharing the same connection between multiple threads, according to the specification at least, however the implementation wants each thread to have its own `IModel`. I believe we can easily get around this using either a write-only connection, or simply putting a lock around the write part of the transport, however i'm not sure how that might affect the performance. 

---
Rebus is [MIT-licensed](https://opensource.org/licenses/MIT). The code submitted in this pull request needs to carry the MIT license too. By leaving this text in, __I hereby acknowledge that the code submitted in the pull request has the MIT license and can be merged with the Rebus codebase__.

### CODE REVIEW

**Comment by zlepper** [comment]
@mookid8000 As agreed, i have done the changes we talked about, can you take a look and see if it looks sane?

File: Rebus.RabbitMq.Tests/RabbitMqHeaderTest.cs [hunk 1]
```diff
@@ -60,7 +60,7 @@ void Callback(object sender, BasicReturnEventArgs eventArgs)
             var bus = StartOneWayClient(Callback);
             await bus.Advanced.Routing.Send(_noneExistingQueueName, "I have headers", headers);
 
-            gotCallback.WaitOrDie(TimeSpan.FromSeconds(2));
+            gotCallback.WaitOrDie(TimeSpan.FromSeconds(5000));
```
Comment (mookid8000) [inline Rebus.RabbitMq.Tests/RabbitMqHeaderTest.cs:63]: Do you really mean to wait for up to ~1.5 hours for this? 😀

File: Rebus.RabbitMq/Internals/ModelObjectPool.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Concurrent;
+using RabbitMQ.Client;
+using Rebus.RabbitMq;
+
+namespace Rebus.Internals
+{
+    /// <summary>
+    /// A simple object pool implementation, because the default (Microsoft.Extensions.ObjectPoo)
```
Comment (mookid8000) [inline Rebus.RabbitMq/Internals/ModelObjectPool.cs:9]:
```
                                                                                                ☝️ 
                                                                                                💩
```

**Review by mookid8000 — APPROVED** [review]

File: Rebus.RabbitMq/Internals/ModelObjectPool.cs (hunk 2 above)
Comment (zlepper) [reply Rebus.RabbitMq/Internals/ModelObjectPool.cs:9]: LOL, good catch XD

File: Rebus.RabbitMq.Tests/RabbitMqHeaderTest.cs (hunk 1 above)
Comment (zlepper) [reply Rebus.RabbitMq.Tests/RabbitMqHeaderTest.cs:63]: No, i was debugging something :P

**Comment by zlepper** [comment]
@mookid8000 Change requests addressed. CI seems to be failing on something unrelated?

**Comment by mookid8000** [comment]
Ok that's weird... looks like it's unable to install RabbitMQ with Chocolatey.... let me just give it a kick

**Comment by zlepper** [comment]
@mookid8000  Build has gone through now. Any chance you can create a beta build? Then we can start more intensive testing internally?

**Comment by mookid8000** [comment]
Sure! It's on NuGet.org now as Rebus.RabbitMq 7.3.1-b4

**Comment by mookid8000** [comment]
Hey @zlepper , I just released Rebus.RabbitMq 7.3.1-b5 where I put some `try...finally` around retrieving/returning models from/to the pool. You probably want to upgrade to that 🙂

**Pull request merged**